substrate/frame/contracts/uapi/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given shared slice.
+	pub fn from_slice(slice: &'a [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+/// Thin-wrapper around a `u32` representing a pointer for Wasm32.
+///
+/// Only for exclusive references.
+///
+/// # Note
+///
+/// Can only be constructed from exclusive reference types and encapsulates the
+/// conversion from reference to raw `u32`.
+/// Does not allow accessing the internal `u32` value.
+#[derive(Debug, Encode)]
+#[repr(transparent)]
+pub struct Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a mut T>,
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given exclusive slice.
+	pub fn from_slice(slice: &'a mut [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: Sized,
+{
+	/// Creates a new Wasm32 pointer from the given exclusive reference.
+	pub fn from_ref(a_ref: &'a mut T) -> Self {
+		let a_ptr: *mut T = a_ref;
+		Self::new(a_ptr as u32)
+	}
+}
+
+/// The raw return code returned by the host side.
+#[repr(transparent)]
+pub struct ReturnCode(u32);
+
+impl From<ReturnCode> for Option<u32> {
+	fn from(code: ReturnCode) -> Self {
+		/// Used as a sentinel value when reading and writing contract memory.
+		///
+		/// We use this value to signal `None` to a contract when only a primitive is
+		/// allowed and we don't want to go through encoding a full Rust type.
+		/// Using `u32::Max` is a safe sentinel because contracts are never
+		/// allowed to use such a large amount of resources. So this value doesn't
+		/// make sense for a memory location or length.
+		const SENTINEL: u32 = u32::MAX;
+
+		(code.0 < SENTINEL).then_some(code.0)
+	}
+}
+
+impl ReturnCode {
+	/// Returns the raw underlying `u32` representation.
+	pub fn into_u32(self) -> u32 {
+		self.0
+	}
+	/// Returns the underlying `u32` converted into `bool`.
+	pub fn into_bool(self) -> bool {
+		self.0.ne(&0)
+	}
+}
+
+type Result = core::result::Result<(), Error>;
+
+#[cfg(not(feature = "std"))]
+#[inline(always)]
+fn extract_from_slice(output: &mut &mut [u8], new_len: usize) {
+	debug_assert!(new_len <= output.len());
+	let tmp = core::mem::take(output);
+	*output = &mut tmp[..new_len];
+}
+
+// is not recognizing its allocator and panic handler definitions.
+#[cfg(not(any(feature = "std", feature = "no-allocator")))]
+mod allocator;
```

Comment:
We neither need an allocator here nor should it be provided by this library. ink! brings its own. Our test fixtures will not need allocation.